In [1]:
import sklearn

In [2]:
!python -m pip install sklearn

In [3]:
%load_ext autoreload

%autoreload 2

In [4]:
import pandas as pd

In [5]:
!dir .\files\auto-mpg.csv

 C 드라이브의 볼륨에는 이름이 없습니다.
 볼륨 일련 번호: 5E33-8C77

 C:\Develops\test_machinelearning\files 디렉터리

2021-06-29  오후 04:04            21,913 auto-mpg.csv
               1개 파일              21,913 바이트
               0개 디렉터리  25,696,727,040 바이트 남음


In [6]:
pd_data = pd.read_csv('./files/auto-mpg.csv', header=None)
pd_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       398 non-null    float64
 1   1       398 non-null    int64  
 2   2       398 non-null    float64
 3   3       398 non-null    object 
 4   4       398 non-null    float64
 5   5       398 non-null    float64
 6   6       398 non-null    int64  
 7   7       398 non-null    int64  
 8   8       398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


In [7]:
pd_data.columns = ['mpg','cylinders','displacement','horsepower','weight',
              'acceleration','model year','origin','name']

In [8]:
pd_data.shape
pd_data[['mpg','weight']]

mpg  weight
0    18.0  3504.0
1    15.0  3693.0
2    18.0  3436.0
3    16.0  3433.0
4    17.0  3449.0
..    ...     ...
393  27.0  2790.0
394  44.0  2130.0
395  32.0  2295.0
396  28.0  2625.0
397  31.0  2720.0

[398 rows x 2 columns]

원 파일에 헤더가 없으므로 dataFrame을 만들면서 header=None으로 헤더가 없다는 것을 알려줌

pd_data.info() -> 결측치 확인

Dtype이 object인 것은 문자도 숫자도 될 수 있으므로 확인 (3,8 column 제외)

pd_data의 column명 지정

x축은 weight y축은 mpg 지정

In [9]:
x = pd_data[['weight']]

In [10]:
y = pd_data[['mpg']]

In [11]:
x.shape, y.shape

((398, 1), (398, 1))

split 기능을 가져옴 (두 집단으로 나눠서 하나는 식을 만들고 나머지는 만들어진 식에 대입하여 확인)

In [12]:
from sklearn.model_selection import train_test_split

In [13]:
X_train, X_test, Y_train, Y_test = train_test_split(x,y)
X_train.shape, X_test.shape, Y_train.shape, Y_test.shape

((298, 1), (100, 1), (298, 1), (100, 1))

sklearn.linear_model에서 선형회귀 기능을 import

In [14]:
from sklearn.linear_model import LinearRegression

In [15]:
lr = LinearRegression()

In [16]:
lr.fit(X_train, Y_train)

LinearRegression()

lr에 직선방정식

In [17]:
# total data
lr.fit(x,y)

LinearRegression()

In [18]:
# split data with 1 column
lr.coef_, lr.intercept_

(array([[-0.00767661]]), array([46.31736442]))

x의 계수

In [19]:
lr.coef_

array([[-0.00767661]])

y 절편

In [20]:
lr.intercept_

array([46.31736442])

weight * mpg의 선형식 : y = - 0.00767661x + 46.31736442

선형식이 얼마나 정확한지 확인하는 방법 : score (정확도)

In [21]:
lr.score(x,y)

0.6917929800341573